### ISSUE TITLE
Questions about experiment details

### ISSUE THREAD

**Issue by speedcell4** [sustcsonglin/TN-PCFG#3]
Hi, thank you for sharing the source code. I got several questions about the data processing details.

1. About the train datasets' sentence length. I found that you set the maximal length of training sentences to be 40 in most of your settings, but only in `fast_tnpcfg_r1000_nt9000_t4500_curriculum0.yaml`, it turns out to be 400. Is this on purpose? Or just a typo?

https://github.com/sustcsonglin/TN-PCFG/blob/7047645f874dcf872ed550d6bcd8d5d2b113d50c/config/fast_tnpcfg_r1000_nt9000_t4500_curriculum0.yaml#L32

2. About dropping sentences with lengths longer than 40, could you please explain the reason for doing this? To avoid out-of-memory? Or just long sentences hurt performance?

3. But you do use all sentences in the dev and test splits without dropping, is this correct?

**Comment by speedcell4** [comment]
Thank you for your quick response.

> I think now it is free to train on arbitrary length sentences, as current implementation is optimized.

Could you please briefly describe what optimization has been applied to your current implementation? I noticed you re-implemented the inside function with `trion`. Is there anything else important?

**Comment by sustcsonglin** [comment]
mainly because of the `log-einsum-exp trick`.  You can also refer to my [pure pytorch implementation](https://github.com/sustcsonglin/TN-PCFG/blob/main/parser/pcfgs/tdpcfg.py#L121C1-L236C14) to get the idea of this trick. In my triton implementation, I do kernel fusion to reduce IO cost in the spirit of FlashAttention.

**Comment by speedcell4** [comment]
Sorry, I don't get it. What do you mean by `log-einsum-exp trick`? is that the [log-sum-exp trick](https://gregorygundersen.com/blog/2020/02/09/log-sum-exp/)? To my knowledge, this is for avoiding overflow, not for accelerating. Could you please describe more details on this?

**Comment by speedcell4** [comment]
Thank you very much~

**Comment by speedcell4** [comment]
Hi, I got one more question.

`log-einsum-exp trick` works fascinatingly well, but is there any way to extend it to the Max semiring for inference?

**Comment by sustcsonglin** [comment]
Pytorch does not have optimized CUDA kernel for generalized matrix multiplication other than `matmul`, so I leverage autograd to estimate marginals and do a simple CKY using the marginals for decoding, namely `MBR decoding` described in the paper.  You might be interested in https://github.com/harvardnlp/genbmm (caveat: a little bit slow), and https://www.kernel-operations.io/keops/index.html ( I did not use it before but you can have a try)

**Comment by speedcell4** [comment]
A little bit different from your answer, for a given distribution `dist`, e.g., CRF, CKY, does the `dist.marginal.argmax(dim=-1)` always equal to `dist.argmax`? I guess the answer is no?

**Comment by sustcsonglin** [comment]
> A little bit different from your answer, for a given distribution `dist`, e.g., labeled CRF, labeled CKY, does the `dist.marginal.argmax(dim=-1)` always equal to `dist.argmax`? I guess the answer is no?

No

**Comment by speedcell4** [comment]
Thanks~
